src/tools/rados/PoolDump.cc

```diff
@@ -22,6 +22,126 @@ using namespace librados;
 #define dout_context g_ceph_context
 #define dout_subsys ceph_subsys_rados
 
+
+
+int PoolDump::export_object(IoCtx *io_ctx, std::string oid, std::string nspace, std::string oloc)
+{
+  ceph_assert(io_ctx != NULL);
+  write_super();
+  dout(10) << __func__ <<  "OID '" << oid << "'" << dendl;
+  return dump_object(io_ctx, oid, nspace, oloc);
+}
+
+int PoolDump::dump_object(IoCtx *io_ctx, std::string oid, std::string nspace, std::string oloc) 
+{
+  // Compose OBJECT_BEGIN
+  // ====================
+  object_begin obj_begin;
+  obj_begin.hoid.hobj.oid = oid;
+  obj_begin.hoid.hobj.nspace = nspace;
+  obj_begin.hoid.hobj.set_key(oloc);
+
+  // Only output head, RadosImport only wants that
+  obj_begin.hoid.hobj.snap = CEPH_NOSNAP;
+
+  // Skip setting object_begin.oi, RadosImport doesn't care
+
+  int r = write_section(TYPE_OBJECT_BEGIN, obj_begin, file_fd);
+  if (r != 0) {
+    return r;
+  }
+
+  // Compose TYPE_DATA chunks
+  // ========================
+  const uint32_t op_size = 4096 * 1024;
+  uint64_t offset = 0;
+  io_ctx->set_namespace(nspace);
+  while (true) {
+    bufferlist outdata;
+    r = io_ctx->read(oid, outdata, op_size, offset);
+    if (r <= 0) {
+      // Error or no data
+      break;
+    }
+
+    r = write_section(TYPE_DATA,
+        data_section(offset, outdata.length(), outdata), file_fd);
+    if (r != 0) {
+      // Output stream error
+      return r;
+    }
+
+    if (outdata.length() < op_size) {
+      // No more data
+      break;
+    }
+    offset += outdata.length();
+  }
+
+  // Compose TYPE_ATTRS chunk
+  // ========================
+  std::map<std::string, bufferlist> raw_xattrs;
+  std::map<std::string, bufferlist> xattrs;
+  r = io_ctx->getxattrs(oid, raw_xattrs);
+  if (r < 0) {
+    cerr << "error getting xattr set " << oid << ": " << cpp_strerror(r)
+         << std::endl;
+    return r;
+  }
+  // Prepend "_" to mimic how user keys are represented in a pg export
+  for (std::map<std::string, bufferlist>::iterator i = raw_xattrs.begin();
```

Comment:
might want to take this opportunity to use range-based loop and structured binding to refactor this `for` loop.